### ISSUE TITLE
Scenario with very slow function execution in Wasmer on Windows compared to other runtimes and a native app

### ISSUE THREAD

**Issue by dsherret** [wasmerio/wasmer#2260]
### Describe the bug

Usually wasmer is pretty quick, but I have a scenario where it's way way slower than browsers and a native rust app.

* wasmer 1.0.2 (cranelift)
* rustc 1.51.0 (2fd73fabe 2021-03-23)
* x86_64
* Edit: Windows 10 (19042.928) -- I haven't been able to reproduce on linux or mac.

### Steps to reproduce

Reproduction steps available in https://github.com/dsherret/slow-wasm-example

### Expected behavior

Would at least expect it to execute somewhere close to chrome (I would hope it would be close to wasmtime though).

### Actual behavior

On my machine:

* Native - ~700ms
* Wasmer (cranelift) - 145s
* Wasmtime - 711ms
* Chrome - 5s
* Firefox - 705ms

This may be because the code is doing many allocations via bumpalo, but usually it's pretty quick...

**Comment by k-nasa** [comment]
I couldn't reproduce it in my environment...

## Enviroment

```
❯ echo "`wasmer -V` | `rustc -V` | `uname -m`"
wasmer 1.0.2 | rustc 1.53.0-nightly (42816d61e 2021-04-24) | x86_64
```

## Results

- native 667ms
- wasmtime 655ms
- wasmer 969ms
- chorme 1539ms
- firefox 686ms

<details>

### native
```
~/lab/oss/sandbox/slow-wasm-example/native-app main*
❯ cargo run --release
    Finished release [optimized] target(s) in 0.01s
     Running `target/release/native-app`
1317605
Finished in 667ms...
```

### wasmtime

```
~/lab/oss/sandbox/slow-wasm-example/wasmtime-app main*
❯ cargo run --release
    Finished release [optimized] target(s) in 0.08s
     Running `target/release/wasmtime-app`
Finished in 655ms...
Result: 1317605
```

### wasmer
```
~/lab/oss/sandbox/slow-wasm-example/wasmer-app main*
❯ cargo run --release
    Finished release [optimized] target(s) in 0.07s
     Running `target/release/app`
Starting...
Finished in 969ms...
Result: I32(1317605)
```

### Chrome

<img width="187" alt="スクリーンショット 2021-04-27 8 51 35" src="https://user-images.githubusercontent.com/23740172/116164995-d2ff8b00-a735-11eb-8c0b-dc98c5b157ff.png">


### Firefox

<img width="165" alt="スクリーンショット 2021-04-27 8 51 01" src="https://user-images.githubusercontent.com/23740172/116165007-d98e0280-a735-11eb-95d2-2f102d631d4f.png">


</details>

**Comment by dsherret** [comment]
@k-nasa interesting! Maybe it only occurs on Windows (I will try out on other OS later). I'm on Windows 10 (19042.928)

**Comment by dsherret** [comment]
Yes, it only seems to occur on windows. I just tried on linux, mac, and two different windows machines. I've updated the description and title.

**Comment by k-nasa** [comment]
I want to try this problem, but I don't have windows machine...

**Comment by syrusakbary** [comment]
I believe the latest PR refactoring traps (#2216) will fix the issue. Although we need to merge #2250 (in progress) #2272 (in progress) first.

I'll keep the issue updated with the progress!

**Comment by ptitSeb** [comment]
I reproduce the issue, even with current wasmer codebase.

I made some profiling, and it seems the cause is an excessive amount of `memcpy` beeing called. Mostly from `core::intrinsics::copy_nonoverlapping<u8>`
Those calls seems to be done from the JIT'd code for the most part, with memory_grow beeing the culprit

![image](https://user-images.githubusercontent.com/3250792/140737809-c997015e-e76d-4faf-96d0-fb7f286cd2cd.png)

**Comment by syrusakbary** [comment]
I believe this might be causing the issue:
https://github.com/wasmerio/wasmer/blob/master/lib/api/src/sys/tunables.rs#L53-L68

**Comment by dsherret** [comment]
Thanks a lot! I can confirm it is fixed and super fast now in 2.1.
